### PR TITLE
Add AlphaTy, BetaTy and GammaTy

### LINKED ISSUE #215: Fix ty with idris backend when using references and parameterised types
we can't to dthis

```
TList : TDef 1
TList = RRef 0

instanceLists : (Ty [List] TList) Nat
instanceList = [1,2,3]
```
because `Ty` expects a vector of `Type` and `List` is of type `Type -> Type`

### PR DESCRIPTION
**Pull request by andrevidela**
fix #215 

Those functions allow to either partially apply a TDef with its
Idris Type representation or use a type constructor to fill one of
its holes.

### CODE REVIEW

File: src/Typedefs/Typedefs.idr [hunk 1]
```diff
@@ -218,6 +218,45 @@ mutual
...
+ApplyVect (x :: xs) c {n = S k} = ApplyVect xs (c x)
+
+ReverseVect : (mkType : Vect n Type -> Type) -> TypeConstructor n
+ReverseVect mkType {n = Z} = mkType []
+ReverseVect mkType {n = (S k)} = \arg => ReverseVect (popHead mkType arg)
+  where
+    popHead : (Vect (S n) Type -> Type) -> Type -> Vect n Type -> Type
+    popHead f t ts = f (t :: ts)
+
+AlphaTy : TDefR n -> TypeConstructor n
```
Comment (clayrat) [inline src/Typedefs/Typedefs.idr:236]: Should we add a docstring here as well?

File: src/Typedefs/Typedefs.idr [hunk 2]
```diff
@@ -218,6 +218,45 @@ mutual
...
+||| Given a vector of type constructors and a TDef construct the Idris type
+||| from it using the second vector to instanciate the constructors
+BetaTy : (types : Vect n (k ** TypeConstructor k)) -> TDefR n
+     -> Vect (NatSum types) Type -> Type
+BetaTy types tdef args = Ty (constructTypes types args) tdef
+
+||| Given a list of type constructors fill the holes with them and return a new type constructor
+GammaTy : (types : Vect n (k ** TypeConstructor k)) -> TDefR n
+     -> TypeConstructor (NatSum types)
+GammaTy types tdef {n} = ReverseVect {n=NatSum types} (BetaTy types tdef)
```
Comment (clayrat) [inline src/Typedefs/Typedefs.idr:258]: Doesn't look like `{n}` is needed here.

**Comment by andrevidela** [comment]
@clayrat I feel like we're getting very close to that since I was looking at how to implement application for type constructor, worth mentionning during the meeting.

File: src/Typedefs/Typedefs.idr (hunk 2 above)
Comment (andrevidela) [reply src/Typedefs/Typedefs.idr]: Fixed

File: src/Typedefs/Typedefs.idr (hunk 1 above)
Comment (andrevidela) [reply src/Typedefs/Typedefs.idr:237]: Added, thanks!

**Pull request merged**